### DISCUSSION TITLE
How to incrementally train a model with spacy v3?

### DISCUSSION THREAD

**Discussion by source19069** [explosion/spaCy#7231] [Help: Coding & Implementations] [answered]
Hello

I have successfully trained a model from a single collection of Docs in a serialized DocBin. I would like to update that initial model with additional Doc collections as we produce them to strengthen the model. They all contain the same kind of annotations.

I use a spacy train command like this for the initial training :
python -m spacy train -V config_initial_train.cfg --paths.train ./PDL1_Value_zzz.spacy  --paths.dev ./PDL1_Value_zzz.spacy --output ./PDL1_Value_train

For resuming training, I supply the path to the  initial model (./PDL1_Value_train) in the config as a 'source' and then run this command, which omits the --output
python -m spacy train -V config_resume_train_cfg --paths.train ./PDL1_Value_yyy.spacy  --paths.dev ./PDL1_Value.spacy_yyy 

This is a straightforward NER model.

spacy appears to be working according to the few examples I've found: for the initial model, it clears the directory to build the new model; upon resuming training, it does not.

However, the size of the model never changes - it's always 3744K.

My questions are:

- Should the model size be changing?

- How can I verify that the model has incorporated all the training data?

Also, I have only used spacy v3, so I have no intuitive knowledge carryover from spacy v2. Please give a more detailed answer that does not assume any spacy v2 experience. The documentation has only scant examples and not much context and I have already reviewed much of the spacy code, so this really an unknown for me.

Here are the config file contents for both the initial and resume training functions I have commented out what I don't need. :
[spacy_config.zip](https://github.com/explosion/spaCy/files/6058161/spacy_config.zip)

## Your Environment
$ python -m spacy info --markdown

## Info about spaCy

- **spaCy version:** 3.0.1
- **Platform:** Windows-2012ServerR2-6.3.9600-SP0
- **Python version:** 3.7.6
- **Pipelines:** en_core_web_sm (3.0.0)

**Comment by honnibal** [comment]
(Ugh sorry I was supposed to comment "moving this to a discussion" on your issue)

The model won't necessarily change size while training on new data. New vocabulary items can be learned with the same number of weights, as we train a variable sized vocabulary using fixed-size embedding tables via the hashing trick.

My suggestion would be to train and test with a relatively small annotations file to see if the model is improving in accuracy on those examples.

**Comment by adrianeboyd** [comment, marked as answer]
What Matt said isn't wrong, but I think the problem here is that you're omitting `--output` in the second train command. The `source` model is only read in from this directory, not updated in place, so your second train command isn't saving the updated model to disk. You want to provide a new `--output` for the second train command.

**Comment by source19069** [comment]
Yes, I see. So the model produced from each training resumption must become the input to the next to achieve the cumulative model. It was not clear if the model was update-in-place or not. Thanks for your response!
